lib/spack/spack/util/url.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+"""
+Utility functions for parsing, formatting, and manipulating URLs.
+"""
```

Comment:
I ended up doing a lot of custom URL parsing/manipulation.  I also noticed other places in the code that would be simpler with utilities for these operations.  So, I added a few in this new module.